File: app/src/main/java/com/android/oleksandrpriadko/demo/loggalitic/DemoPublisher.java
package com.android.oleksandrpriadko.demo.loggalitic;

import java.util.Map;

import android.os.Bundle;
import android.support.v4.util.Pair;
import android.util.Log;

import com.android.oleksandrpriadko.demo.App;
import com.android.oleksandrpriadko.loggalitic.analytics.AnalyticsEvent;
import com.android.oleksandrpriadko.loggalitic.analytics.Publisher;
import com.android.oleksandrpriadko.loggalitic.policy.Policy;
import com.crashlytics.android.answers.Answers;
import com.microsoft.appcenter.analytics.Analytics;

public class DemoPublisher extends Publisher {

    public DemoPublisher(Policy policy) {
        super(policy);
    }

    @Override
    public boolean send(AnalyticsEvent event) {
        return toFabric(event) && toFireBase(event) && toAppCenter(event);
    }

    private boolean toFabric(AnalyticsEvent event) {
        FabricConverter fabricConverter = findConverter(FabricConverter.class);
        if (fabricConverter != null) {
            Answers.getInstance().logCustom(fabricConverter.convert(event));
            return true;
        } else {
            Log.d(getTag(),
                "send: failed. reason = " + FabricConverter.class.getSimpleName() + " is null");
            return false;
        }
    }

    private boolean toFireBase(AnalyticsEvent event) {
        FireBaseConverter fireBaseConverter = findConverter(FireBaseConverter.class);
        if (fireBaseConverter != null) {
            Pair<String, Bundle> pair = fireBaseConverter.convert(event);
            App.getFireBaseAnalytics().logEvent(pair.first, pair.second);
            return true;
        } else {
            Log.d(getTag(),
                "send: failed. reason = " + FireBaseConverter.class.getSimpleName() + " is null");
            return false;
        }
    }

    private boolean toAppCenter(AnalyticsEvent event) {
        AppCenterConverter appCenterConverter = findConverter(AppCenterConverter.class);
        if (appCenterConverter != null) {
            Pair<String, Map<String, String>> pair = appCenterConverter.convert(event);
            Analytics.trackEvent(pair.first, pair.second);
            return true;
        } else {
            Log.d(getTag(),
                "send: failed. reason = " + AppCenterConverter.class.getSimpleName() + " is null");
            return false;
        }
    }
}

File: app/src/main/java/com/android/oleksandrpriadko/demo/logpublish/DemoPublisher.kt
package com.android.oleksandrpriadko.demo.logpublish

import android.util.Log
import com.android.oleksandrpriadko.core.policy.AnalyticsEvent
import com.android.oleksandrpriadko.core.policy.Policy
import com.android.oleksandrpriadko.demo.main.App
import com.android.oleksandrpriadko.loggalitic.LogPublishService
import com.android.oleksandrpriadko.loggalitic.analytics.Publisher
import com.crashlytics.android.answers.Answers
import com.microsoft.appcenter.analytics.Analytics
import io.fabric.sdk.android.Fabric

class DemoPublisher(policy: Policy) : Publisher(policy) {

    override fun send(event: AnalyticsEvent): Boolean {
        return try {
            toFabric(event) && toFireBase(event) && toAppCenter(event)
        } catch (ignore: Exception) {
            false
        }
    }

    private fun toFabric(event: AnalyticsEvent): Boolean {
        val fabricConverter = findConverter<FabricConverter>(FabricConverter::class.java)
        return when {
            fabricConverter != null -> {
                try {
                    Answers.getInstance().logCustom(fabricConverter.convert(event))
                } catch (e: IllegalStateException) {
                    LogPublishService.logger().d(tag,
                            "send: failed. reason = ${Fabric::class.java.simpleName} not initialised")
                    return false
                }
                true
            }
            else -> {
                LogPublishService.logger().d(tag,
                        "send: failed. reason = ${FabricConverter::class.java.simpleName} is null")
                false
            }
        }
    }

    private fun toFireBase(event: AnalyticsEvent): Boolean {
        val fireBaseConverter = findConverter<FireBaseConverter>(FireBaseConverter::class.java)
        return if (fireBaseConverter != null) {
            val pair = fireBaseConverter.convert(event)
            App.fireBaseAnalytics!!.logEvent(pair.first!!, pair.second)
            true
        } else {
            Log.d(tag,
                    "send: failed. reason = ${FireBaseConverter::class.java.simpleName} is null")
            false
        }
    }

    private fun toAppCenter(event: AnalyticsEvent): Boolean {
        val appCenterConverter = findConverter<AppCenterConverter>(AppCenterConverter::class.java)
        return if (appCenterConverter != null) {
            val pair = appCenterConverter.convert(event)
            Analytics.trackEvent(pair.first, pair.second)
            true
        } else {
            LogPublishService.logger().d(tag,
                    "send: failed. reason = ${AppCenterConverter::class.java.simpleName} is null")
            false
        }
    }
}